Download JDK and create distribution artifacts with JDK

andsel opened the pull request:
This PR add the bundling of JDK to existing artifacts tasks
The artifacts are:
- Linux distribution's packages: `deb/rpm`
- normal packages `tar.gz` (Linux, MacOS) `.zip` for windows
- Docker images

The JDK to bundle is configured in `versions.yml` file, and are the same for all platform (AdoptOpenJDK 11)

The launch scripts has changed with the following meaning, in case of a bundled package:
use the JVM provided by JAVA_HOME if set, warning the user that he is using a bundled package. If not set use the bundled one if present. As last resource (only on *nix platforms) check for other java with `command -v java`, if no JVM is present quit with error.

On DEB package with JDK bundled removes the dependency on a Java DEB.

----
This PR bring an effort from @robbavey  and proposes changes to adapt current Logstash packaging scripts to create artifacts containing the JDK. 
- [x] download JDK to bundle from an Elastic repository, (Adopt for Linux and Zulu for Windows)
- [x] bundle the JDK with tar.gz distribution (keeping also a distribution without it)
- [x] change launching scripts in `bin/`to use the shipped JDK if present
- [x] add deb/rpm with also bundled JDK, in this case the dependency on Java should be eliminated (because it bundles one JDK)
- [x] align the JDK selection to what does elasticsearch, if JAVA_HOME is already set use that, else use the bundled one without changing the JAVA_HOME

andsel commented on logstash-core/build.gradle:
```diff
@@ -68,8 +68,10 @@ task copyGemjar(type: Copy, dependsOn: [sourcesJar, copyRuntimeLibs]) {
 task cleanGemjar {
-    delete fileTree(project.file('lib/jars/')) {
-        include '*.jar'
+    doLast {
```
Note for reviewer:
if the deletion is kept in Gradle's configuration phase, then when we run any other task on the parent project, this piece is executed, resulting in undesired deletions. This happens because Gradle parent project is configured, so configure the subprojects and configuring `logstash-core` scatter the deletion. If I invoked the copyJdk to prepare for bundle it ends in a tar without the jars that composes Logstash core (logstash-core.jar)

andsel commented:
Jenkins test this please

robbavey commented on versions.yml:
```diff
@@ -4,6 +4,20 @@ logstash: 8.0.0
+bundled_jdk:
+  linux:
+      vendor: "adoptopenjdk"
+      revision: 14
```
Are we wanting to bundle JDK 14? Or stick to LTS releases?

andsel replied:
we decided for LTS (11 at the moment)

robbavey commented on versions.yml:
```diff
@@ -4,6 +4,20 @@ logstash: 8.0.0
+      build: 36
+  windows:
+      vendor: "zulu"
+      revision: 11.39.15
```
If we do choose to bundle Zulu instead of adopt open jdk we would need to keep the major versions the same

andsel replied:
We decided to use AdoptOpenJDK also for windows

andsel commented on bin/system-install:
```diff
@@ -61,7 +61,7 @@ done
 # bin/logstash-plugin is a short lived ruby script thus we can use aggressive "faster starting JRuby options"
 # see https://github.com/jruby/jruby/wiki/Improving-startup-time
-export JRUBY_OPTS="$JRUBY_OPTS -J-XX:+TieredCompilation -J-XX:TieredStopAtLevel=1 -J-noverify -X-C -Xcompile.invokedynamic=false"
+export JRUBY_OPTS="$JRUBY_OPTS -J--add-opens java.base/sun.nio.ch=org.jruby.dist -J-XX:+TieredCompilation -J-XX:TieredStopAtLevel=1 -J-noverify -X-C -Xcompile.invokedynamic=false"
```
You are right, fixed with 1fb3db47b3ff51bd7476c5f5b2e619e4b7c679f1

andsel replied:
This flag has been removed since those `--add-opens` comes from JRuby specific file: fixed with 9b9b2600238ceefd4a48546c7fd7d828163b30c8

andsel commented on build.gradle:
```diff
@@ -510,10 +512,138 @@ bootstrap.dependsOn assemblyDeps
-Boolean oss = System.getenv('OSS').equals('true')
+    String createDownloadUrl() {
+        if (vendor == "adoptopenjdk") {
+            return "https://artifactory.elstc.co/artifactory/oss-jdk-local/adoptopenjdk/OpenJDK${major}U-jdk_x64_${osName}_hotspot_${revision}_${build}.${extension}"
```
From the repo https://artifactory.elstc.co/artifactory/oss-jdk-local/adoptopenjdk/ I see that `aarch64` is only available for Linux, so should we add a sanity check when a we define also the architecture in `versions.yml`? (es check that os is Linux if cpu architecture is `aarch64`

andsel commented on versions.yml:
```diff
@@ -4,6 +4,20 @@ logstash: 8.0.0
+bundled_jdk:
+  linux:
+      vendor: "adoptopenjdk"
+      revision: 13.0.2
```
we mirror what ES does (AdoptOpenJDK latest for all OSes, 8f7db5ce4bd468d8f6743c3351b700ae4a08519a)

andsel commented on build.gradle:
```diff
@@ -438,9 +440,138 @@ task copyEs(type: Copy, dependsOn: [downloadEs, deleteLocalEs]) {
+        if (vendor == "adoptopenjdk") {
+            return "https://artifactory.elstc.co/artifactory/oss-jdk-local/adoptopenjdk/OpenJDK${major}U-jdk_x64_${osName}_hotspot_${revision}_${build}.${extension}"
+        }
+        if (vendor == "zulu") {
```
Download from AdoptOpenJDK site the version 11 fix: 1dd27a1965d74bbbc088d1f1c03be0876d3fccc0

andsel commented on build.gradle:
```diff
@@ -509,11 +511,152 @@ bootstrap.dependsOn assemblyDeps
+}
 
-if (!oss) {
+if (System.getenv('OSS') != 'true') {
```
Note for reviewer, `oss` wasn't defined

andsel commented on rakelib/artifacts.rake:
```diff
@@ -59,9 +59,12 @@ namespace "artifact" do
       # plugins which use jar-dependencies.
       # See more in https://github.com/elastic/logstash/issues/4818
       "vendor/??*/**/.mvn/**/*",
+      "vendor/jruby/bin/.jruby.java_opts",
```
Note for reviewer: without this on bundled AdoptOpenJDK the JRuby execution used to run `pleaserun` in post install script is run without the opening of modules for private introspection (miss --add-opens ...)

andsel commented:
Hi @robbavey it's ready for review, I've fixed the Mac part but I don't have a Mac to test run of the untarred package (with bundled jdk)

robbavey commented:
Jenkins test this please

andsel commented on build.gradle:
```diff
@@ -509,11 +511,152 @@ bootstrap.dependsOn assemblyDeps
+                        "jdk-${revision}+${build}":
+                        "jdk${revision}u${build}"
+                return "https://api.adoptopenjdk.net/v3/binary/version/${releaseName}/${osName}/${arch}/jdk/hotspot/normal/${vendor}"
+            case "zulu":
```
Actually we bundle AdoptOpenJDK on Windows, but if we switch / want to switch to Zulu the stuff is already in place

andsel commented on versions.yml:
```diff
@@ -4,6 +4,21 @@ logstash: 8.0.0
 logstash-core: 8.0.0
 logstash-core-plugin-api: 2.1.16
 
+bundled_jdk:
```
You are right, was done in respect to differentiate JVM per OS, we could keep open this availability while simplifying, I mean we could have something like:
```
bundled_jdk:
  vendor: "adoptopenjdk"
  revision: 11.0.8
  build: 10
```
and this is valid for every OS, then if we want to customize for Windows we could have a subsection only for that customization: something like:
```
bundled_jdk:
  vendor: "adoptopenjdk"
  revision: 11.0.8
  build: 10

 windows:
   vendor: "zulu"
   revision: 11.39.15
   build: 0.7
```
In this case Mac and Linux would use AdoptOpenJDK while Windows Zulu. This complicates little bit the Groovy code, but not so much I think
WDYT?

robbavey replied:
I think I'd be inclined to keep it simple until we need it - if the requirements change and we need to have different java distributions per platform we can add the extra code, but leave it until then

andsel replied:
done

andsel commented on bin/logstash.lib.sh:
```diff
@@ -75,12 +75,28 @@ parse_jvm_options() {
   if [ -z "$JAVACMD" ]; then
-    if [ -x "$JAVA_HOME/bin/java" ]; then
-      JAVACMD="$JAVA_HOME/bin/java"
+    JAVACMD_TEST=`command -v java`
```
that if it has been moved below https://github.com/elastic/logstash/pull/11845/files/6d36538c2bd2f94634cf8137723bb16e5784aa2d#diff-153c0379c26855674ab54f42c924c7efR93.

The flow should be if no JAVA_HOME and no Java are present in the system use the bundled (determining the correct path in case of Darwin) else detect if it's present JAVA_HOME or the java command and use that

andsel replied:
Chatting with @yaauie, he came out with an idea. We could use a different priority, so first check `JAVA_HOME`, if not exists use the bundled JDK (if present) then fallback to `command -v java` to check if any Java is installed, else raise an error message.
In case of bundled JDK if the user is defining a JAVA_HOME we could log a message that he is using a Logstash bundled with JDK and use a customized JAVA_HOME, so warning him that probably he could use a Logstash distribution without  bundled JDK. WDYT?

robbavey replied:
Yeah, that makes sense to me. It makes the choice of using the bundled Java or a custom Java relatively straightforward and logging the fact that the running Java is sourced via `JAVA_HOME` helps surface users potentially inadvertently using a custom Java.

andsel replied:
done

andsel commented:
Jenkins test this please

andsel commented:
@robbavey  about the the question "how to test?" we could think to extend the `ci/acceptance_tests.sh` to run the installation of deb/rpm packages inside the VMs with Vagrant, but we have the limit that the deb/rpm creation must run on a JDK8 due to the limits of `fpm`.

andsel commented:
Hi @robbavey, fixed failing test was due nesting call of gradle that invoke rake that invoke gradle tasks.

andsel commented on build.gradle:
```diff
@@ -509,11 +511,152 @@ bootstrap.dependsOn assemblyDeps
+    String createDownloadUrl() {
+        switch (vendor) {
+            case "adoptopenjdk":
+            case "openjdk":
```
I followed the [swagger docs](https://api.adoptopenjdk.net/swagger-ui/#/Binary/get_v3_binary_version__release_name___os___arch___image_type___jvm_impl___heap_size___vendor_) but trying the url https://api.adoptopenjdk.net/v3/binary/version/jdk-11.0.6%2B10/linux/x64/jdk/hotspot/normal/openjdk seems it's broken

andsel commented on build.gradle:
```diff
@@ -509,11 +511,152 @@ bootstrap.dependsOn assemblyDeps
+    outputs.dir(project.ext.jdkBundlingDirectory)
+}
+
+task unzipJdk(type: Copy, dependsOn: downloadJdk) {
```
This was failing because the `deleteLocalJdk` task doesn't execute anything due to missed `osType` parameter, now added and shouldn't happen

andsel commented on rakelib/artifacts.rake:
```diff
@@ -123,53 +126,93 @@ namespace "artifact" do
+        build_tar('ELASTIC-LICENSE', platform: '-linux-x86-64')
+      when "windows"
+        build_zip('ELASTIC-LICENSE', platform: '-windows-x86-64')
+      when "mac"
```
done

andsel commented on rakelib/artifacts.rake:
```diff
@@ -515,8 +564,12 @@ namespace "artifact" do
     out.name = oss ? "logstash-oss" : "logstash"
-    out.version = "#{LOGSTASH_VERSION}#{PACKAGE_SUFFIX}".gsub(/[.-]([[:alpha:]])/, '~\1')
     out.architecture = "all"
+    if bundle_jdk
```
done

andsel commented on rakelib/artifacts.rake:
```diff
@@ -123,53 +126,93 @@ namespace "artifact" do
-    build_tar('APACHE-LICENSE-2.0', "-oss", oss_excluder)
-  end
+  desc "Build a tar.gz and zip of default logstash plugins with all dependencies"
+  task "archives" => ["prepare", "generate_build_metadata"] do
```
`task "all"` depends on `prepare` and `build`and the task build, which is a meta task used to invoke others task, has been update from:
```
    Rake::Task["artifact:zip"].invoke
    Rake::Task["artifact:zip_oss"].invoke
    Rake::Task["artifact:archives"].invoke
    Rake::Task["artifact:tar_oss"].invoke
```
to

```
    Rake::Task["artifact:archives"].invoke
    Rake::Task["artifact:archives_oss"].invoke
```
with https://github.com/elastic/logstash/pull/11845/files#diff-54706f91d613c53c409a4e3f70c6aebdL238

The `docker` `docker_oss` task dependecy on `tar` and `tar_oss` has been changed to `archives` and `archives_oss` respectively

andsel commented:
Hi @robbavey with latest changes I've added an acceptance to test to verify the usage of bundled JDK and fixed all your suggestions, if you would like to review this, it's ready

andsel commented:
Jenkins test this please

robbavey commented on bin/setup.bat:
```diff
@@ -22,8 +22,16 @@ rem ### 2: set java
   set JAVA="%JAVA_HOME%\bin\java.exe"
+  echo Using JAVA_HOME defined java: %JAVA_HOME%
+  if exist "%LS_HOME%\jdk" (
+    echo WARNING, using JAVA_HOME while Logastash distribution comes with a bundled JDK
```
Nit: Typo 'Logstash'

robbavey commented on bin/logstash.lib.sh:
```diff
@@ -72,12 +72,33 @@ parse_jvm_options() {
+      if [ -x "$JAVA_HOME/bin/java" ]; then
+        JAVACMD="$JAVA_HOME/bin/java"
+        if [ -d "${LOGSTASH_HOME}/${BUNDLED_JDK_PART}" -a -x "${LOGSTASH_HOME}/${BUNDLED_JDK_PART}/bin/java" ]; then
+          echo "WARNING, using JAVA_HOME while Logastash distribution comes with a bundled JDK"
```
Nit: Typo 'Logstash'

robbavey commented on rakelib/artifacts.rake:
```diff
@@ -437,6 +496,12 @@ namespace "artifact" do
     end
 
+    if bundle_jdk
+      suffix += "-with-jdk-x86_64"
```
Is the intention to diverge from elasticsearch and have `-with-jdk-{arch}` and `-no-jdk`, or have `{arch}` and `-no-jdk`?

robbavey commented on rakelib/artifacts.rake:
```diff
@@ -540,13 +607,23 @@ namespace "artifact" do
     # TODO(sissel): Include the git commit hash?
     out.iteration = "1" # what revision?
-    out.url = "http://www.elasticsearch.org/overview/logstash/"
+    out.url = "https://www.elastic.co/logstash"
```
Nice catch!

robbavey commented on rakelib/artifacts.rake:
```diff
@@ -540,13 +607,23 @@ namespace "artifact" do
+    out.url = "https://www.elastic.co/logstash"
     out.description = "An extensible logging pipeline"
-    out.vendor = "Elasticsearch"
+    out.vendor = "Elastic"
```
And another one!

robbavey commented on qa/rspec/commands.rb:
```diff
@@ -114,7 +119,16 @@ def to_s
+        jdk_infix = ""
+      else
+        if @bundled_jdk
+          jdk_infix = "with-jdk-x86_64-"
```
Do we need `with-jdk` here?

andsel commented:
Jenkins test this please

andsel commented:
Jenkins test this please

robbavey commented on bin/logstash.lib.sh:
```diff
@@ -72,12 +72,33 @@ parse_jvm_options() {
 }
 
+setup_bundled_jdk_part() {
+  if [ "$OSTYPE" = "darwin"* ]; then
```
```suggestion
  if [[ "$OSTYPE" = "darwin"* ]]; then
```

Requires double brackets for the wildcard test to work

andsel replied:
The `[[ ... ]]` is a bash syntax, the problem is that `logstash.lib.sh` is included both in `#!/bin/sh` scripts and `#!/bin/bash` so we can't use bash specific idioms. I switched to `uname -s`  and sh string comparison:
```sh
OS_NAME="$(uname -s)"
if [ $OS_NAME = "Darwin" ]; then
 ...
fi
```

robbavey commented on bin/logstash.lib.sh:
```diff
@@ -72,12 +72,33 @@ parse_jvm_options() {
+      JAVACMD="${LOGSTASH_HOME}/${BUNDLED_JDK_PART}/bin/java"
+    elif [ -n "$JAVACMD_TEST" ]; then
       set +e
       JAVACMD=`command -v java`
```
Maybe we should add a log message here to state that we are using the system java, and the location of it.

andsel replied:
Added

robbavey reviewed (changes requested):
Review still in progress, adding comments on breakage on MacOs machines

robbavey reviewed (changes requested):
Getting closer, but I'm still unable to run `rake artifact:all` - the recently added `docker_ubi8` task still relies on `tar`, and fails due to this requirement

Before merging, can we see a full ci test on jenkins, including the acceptance tests across platforms and docker?

andsel commented:
So we can simply change the `docker_ubi8` task to rely on `archives` task that replaces the old `tar` task.
Regarding the Jenkins build, I wasn't able to trigger the acceptance build on this branch

robbavey commented on rakelib/artifacts.rake:
```diff
@@ -396,33 +451,37 @@ namespace "artifact" do
   end
 
-  def package(platform, version, variant=:standard)
+  def package_with_jdk(platform, version, variant=:standard)
```
`version` vs `vendor` - `version` is unused here, instead `vendor` is passed to `package` method

andsel replied:
used version, however it worked because `version` in `package` method never used

robbavey commented on rakelib/artifacts.rake:
```diff
@@ -529,7 +599,9 @@ namespace "artifact" do
         out.attributes[:deb_user] = "root"
         out.attributes[:deb_group] = "root"
-        out.attributes[:deb_suggests] = "java8-runtime-headless"
+        if not bundle_jdk
```
Nit: Maybe `unless` would read better here

robbavey reviewed (changes requested):
Just a few nitpicks

One caveat - I've not yet tested this on a Windows box - have we tested that yet?

andsel commented:
With this PR the install folder contains the arch and OS (for example: `logstash-8.0.0-SNAPSHOT-linux-x86_64`) I thought this was a good idea since after decompress the directory explains well what it contains. In the worst case `-windows-x86_64` we add 15 chars, I forgot to check such a limitation for windows. Do you think it worthwhile?

robbavey commented:
@andsel Yeah, I know - and not having consistency throughout the stack makes the question a little harder. My thought on keeping it the same was that we keep the install directory for logstash the same, regardless of whether it is `oss` or the default distribution, so I was wondering if we should do the same with the architecture too, and what value having the architecture in the directory name gave us.

robbavey reviewed (changes requested):
One more thing - we use `x86-64` in some places, and `x86_64` in others. We should probably use `x86_64` consistently everywhere

andsel commented on docker/templates/Dockerfile.j2:
```diff
@@ -29,7 +29,7 @@ FROM {{ base_image }}
 # scripts.
 # Minimal distributions also require findutils tar gzip (procps for integration tests)
-RUN {{ package_manager }} update -y && {{ package_manager }} install -y procps findutils tar gzip java-11-openjdk-devel which && \
+RUN {{ package_manager }} update -y && {{ package_manager }} install -y procps findutils tar gzip which shadow-utils && \
```
@robbavey shadow-utils bring in `groupadd` used at line 36, probably was a transitive dependency of the `java-11-openjdk-devel`

robbavey replied:
Good catch

robbavey commented on bin/setup.bat:
```diff
@@ -22,8 +22,16 @@ rem ### 2: set java
+  )
 ) else (
-  for %%I in (java.exe) do set JAVA="%%~$PATH:I"
+  if exist "%LS_HOME%\jdk" (
```
Can we add informative messages stating which JDK we are using, like we do with the bash scripts?

robbavey reviewed (commented):
Tested with Windows, and I have a small nit on logging, but other than that, I think we're good to go

robbavey reviewed (approved):
Assuming the acceptance tests are green, LGTM. 

Thank you for patience working through all the changes, I know it has been a long process

andsel commented:
Jenkins test this please
